File: order.cpp
#include "order.hpp"
#include <iostream>
#include <iomanip>
#include <algorithm>
#include <ctime>

// Инициализация статической переменной для генерации ID заказов
int Order::s_nextOrderId = 1;

Order::Order()
	: m_orderId(s_nextOrderId++)
	, m_menuItems()
	, m_totalAmount(0.0)
	, m_orderTime(std::time(nullptr))
	, m_employee(nullptr)
{
}

Order::Order(Employee* employee)
	: m_orderId(s_nextOrderId++)
	, m_menuItems()
	, m_totalAmount(0.0)
	, m_orderTime(std::time(nullptr))
	, m_employee(employee)
{
}

bool Order::AddMenuItem(MenuItem* menuItem)
{
	if (menuItem != nullptr)
	{
		// Проверяем, доступно ли блюдо к продаже
		if (menuItem->IsAvailable())
		{
			m_menuItems.push_back(menuItem);
			CalculateTotalAmount();	// Пересчитываем сумму после добавления
			std::cout << "Блюдо '" << menuItem->GetName() 
					  << "' добавлено в заказ #" << m_orderId << std::endl;
			return true;
		}
		else
		{
			std::cout << "Блюдо '" << menuItem->GetName() 
					  << "' недоступно к продаже!" << std::endl;
			return false;
		}
	}
	return false;
}

bool Order::RemoveMenuItem(const std::string& menuItemName)
{
	auto it = std::find_if(m_menuItems.begin(), m_menuItems.end(),
		[&menuItemName](MenuItem* item) {
			return item != nullptr && item->GetName() == menuItemName;
		});

	if (it != m_menuItems.end())
	{
		m_menuItems.erase(it);
		CalculateTotalAmount();	// Пересчитываем сумму после удаления
		std::cout << "Блюдо '" << menuItemName 
				  << "' удалено из заказа #" << m_orderId << std::endl;
		return true;
	}
	
	std::cout << "Блюдо '" << menuItemName 
			  << "' не найдено в заказе #" << m_orderId << std::endl;
	return false;
}

double Order::CalculateTotalAmount()
{
	m_totalAmount = 0.0;
	
	for (const auto* menuItem : m_menuItems)
	{
		if (menuItem != nullptr)
		{
			m_totalAmount += menuItem->GetSellingPrice();
		}
	}
	
	return m_totalAmount;
}

void Order::PrintOrder() const
{
	std::cout << "=== Заказ #" << m_orderId << " ===" << std::endl;
	
	// Вывод времени заказа
	char timeStr[100];
	std::strftime(timeStr, sizeof(timeStr), "%d.%m.%Y %H:%M:%S", std::localtime(&m_orderTime));
	std::cout << "Время создания: " << timeStr << std::endl;
	
	// Вывод информации о сотруднике
	if (m_employee != nullptr)
	{
		std::cout << "Сотрудник: " << m_employee->GetFullName() 
				  << " (" << m_employee->GetPosition() << ")" << std::endl;
	}
	else
	{
		std::cout << "Сотрудник: не указан" << std::endl;
	}
	
	std::cout << std::endl;
	std::cout << "Блюда в заказе:" << std::endl;
	
	if (m_menuItems.empty())
	{
		std::cout << "  Заказ пуст." << std::endl;
	}
	else
	{
		int index = 1;
		for (const auto* menuItem : m_menuItems)
		{
			if (menuItem != nullptr)
			{
				std::cout << "  " << index << ". " << menuItem->GetName() 
						  << " - " << std::fixed << std::setprecision(2) 
						  << menuItem->GetSellingPrice() << " руб." << std::endl;
				index++;
			}
		}
	}
	
	std::cout << std::endl;
	std::cout << "Итого: " << std::fixed << std::setprecision(2) 
			  << m_totalAmount << " руб." << std::endl;
	std::cout << "==========================" << std::endl;
}

void Order::CompleteOrder()
{
	// Обновляем счетчики продаж для каждого блюда в заказе
	for (auto* menuItem : m_menuItems)
	{
		if (menuItem != nullptr)
		{
			int currentSales = menuItem->GetSalesCount();
			menuItem->SetSalesCount(currentSales + 1);
		}
	}
	
	// Добавляем продажу сотруднику
	if (m_employee != nullptr)
	{
		m_employee->AddSales(static_cast<int>(m_menuItems.size()));
	}
	
	std::cout << "Заказ #" << m_orderId << " завершен. Продажи обновлены." << std::endl;
}

int Order::GetOrderId() const
{
	return m_orderId;
}

Order::MenuItemsType Order::GetMenuItems() const
{
	return m_menuItems;
}

double Order::GetTotalAmount() const
{
	return m_totalAmount;
}

std::time_t Order::GetOrderTime() const
{
	return m_orderTime;
}

Employee* Order::GetEmployee() const
{
	return m_employee;
}

int Order::GetMenuItemsCount() const
{
	return static_cast<int>(m_menuItems.size());
}

void Order::SetEmployee(Employee* employee)
{
	m_employee = employee;
}

File: lab4/Order.java
import java.util.ArrayList;
import java.text.SimpleDateFormat;
import java.util.Date;

// Класс Order реализует интерфейс ICloneable для демонстрации клонирования
public class Order implements ICloneable<Order> {
    // Тип для списка позиций меню
    public static class MenuItemsType extends ArrayList<MenuItem> {
        // Наследуемся от ArrayList для удобства работы
    }

    // Статическая переменная для генерации уникальных ID заказов
    private static int nextOrderId = 1;

    // Приватные поля
    private int orderId;                  // ID заказа
    private MenuItemsType menuItems;      // Блюда в заказе
    private double totalAmount;           // Сумма заказа
    private long orderTime;               // Время создания заказа (в миллисекундах)
    private Employee employee;            // Сотрудник, добавивший заказ

    // Конструктор по умолчанию
    public Order() {
        this.orderId = nextOrderId++;
        this.menuItems = new MenuItemsType();
        this.totalAmount = 0.0;
        this.orderTime = System.currentTimeMillis();
        this.employee = null;
    }

    // Конструктор с параметрами
    public Order(Employee employee) {
        this.orderId = nextOrderId++;
        this.menuItems = new MenuItemsType();
        this.totalAmount = 0.0;
        this.orderTime = System.currentTimeMillis();
        this.employee = employee;
    }

    // Добавить блюдо в заказ
    public boolean addMenuItem(MenuItem menuItem) {
        if (menuItem != null) {
            // Проверяем, доступно ли блюдо к продаже
            if (menuItem.isAvailable()) {
                menuItems.add(menuItem);
                calculateTotalAmount(); // Пересчитываем сумму после добавления
                System.out.println("Блюдо '" + menuItem.getName() + "' добавлено в заказ #" + orderId);
                return true;
            } else {
                System.out.println("Блюдо '" + menuItem.getName() + "' недоступно к продаже!");
                return false;
            }
        }
        return false;
    }

    // Удалить блюдо из заказа
    public boolean removeMenuItem(String menuItemName) {
        for (int i = 0; i < menuItems.size(); i++) {
            MenuItem item = menuItems.get(i);
            if (item != null && item.getName().equals(menuItemName)) {
                menuItems.remove(i);
                calculateTotalAmount(); // Пересчитываем сумму после удаления
                System.out.println("Блюдо '" + menuItemName + "' удалено из заказа #" + orderId);
                return true;
            }
        }
        
        System.out.println("Блюдо '" + menuItemName + "' не найдено в заказе #" + orderId);
        return false;
    }

    // Рассчитать сумму заказа
    public double calculateTotalAmount() {
        totalAmount = 0.0;
        
        for (MenuItem menuItem : menuItems) {
            if (menuItem != null) {
                totalAmount += menuItem.getSellingPrice();
            }
        }
        
        return totalAmount;
    }

    // Вывести информацию о заказе
    public void printOrder() {
        System.out.println("=== Заказ #" + orderId + " ===");
        
        // Вывод времени заказа
        SimpleDateFormat dateFormat = new SimpleDateFormat("dd.MM.yyyy HH:mm:ss");
        System.out.println("Время создания: " + dateFormat.format(new Date(orderTime)));
        
        // Вывод информации о сотруднике
        if (employee != null) {
            System.out.println("Сотрудник: " + employee.getFullName() + " (" + employee.getPosition() + ")");
        } else {
            System.out.println("Сотрудник: не указан");
        }
        
        System.out.println();
        System.out.println("Блюда в заказе:");
        
        if (menuItems.isEmpty()) {
            System.out.println("  Заказ пуст.");
        } else {
            int index = 1;
            for (MenuItem menuItem : menuItems) {
                if (menuItem != null) {
                    System.out.printf("  %d. %s - %.2f руб.%n", index, 
                                     menuItem.getName(), menuItem.getSellingPrice());
                    index++;
                }
            }
        }
        
        System.out.println();
        System.out.printf("Итого: %.2f руб.%n", totalAmount);
        System.out.println("==========================");
    }

    // Метод для завершения заказа (обновляет счетчики продаж)
    public void completeOrder() {
        // Обновляем счетчики продаж для каждого блюда в заказе
        for (MenuItem menuItem : menuItems) {
            if (menuItem != null) {
                // Используем метод incrementSalesCount из MenuItem
                menuItem.incrementSalesCount();
            }
        }
        
        // Добавляем продажу сотруднику
        if (employee != null) {
            employee.addSales(menuItems.size());
        }
        
        System.out.println("Заказ #" + orderId + " завершен. Продажи обновлены.");
    }

    // Геттеры
    public int getOrderId() {
        return orderId;
    }

    public MenuItemsType getMenuItems() {
        return menuItems;
    }

    public double getTotalAmount() {
        return totalAmount;
    }

    public long getOrderTime() {
        return orderTime;
    }

    public Employee getEmployee() {
        return employee;
    }

    public int getMenuItemsCount() {
        return menuItems.size();
    }

    // Сеттеры
    public void setEmployee(Employee employee) {
        this.employee = employee;
    }
    
    // Реализация методов интерфейса ICloneable
    @Override
    public Order cloneShallow() {
        // Поверхностное клонирование: создаем новый Order, но ссылки на MenuItem остаются те же
        Order cloned = new Order();
        cloned.orderId = nextOrderId++;  // Новый ID для клона (используем статическое поле)
        cloned.orderTime = this.orderTime;
        cloned.employee = this.employee;  // Та же ссылка на сотрудника
        cloned.menuItems = this.menuItems;  // Та же ссылка на список (shallow copy)
        cloned.calculateTotalAmount();  // Пересчитываем сумму
        return cloned;
    }
    
    @Override
    public Order cloneDeep() {
        // Глубокое клонирование: создаем новые объекты MenuItem для каждого блюда
        Order cloned = new Order();
        cloned.orderId = nextOrderId++;  // Новый ID для клона (используем статическое поле)
        cloned.orderTime = this.orderTime;
        cloned.employee = this.employee;  // Сотрудник остается тем же (можно было бы клонировать, но оставим как есть)
        
        // Глубокое копирование MenuItem - создаем новые объекты
        cloned.menuItems = new MenuItemsType();
        for (MenuItem item : this.menuItems) {
            if (item != null) {
                // Создаем новый MenuItem с теми же параметрами
                MenuItem clonedItem = new MenuItem(
                    item.getName(),
                    item.getCostPrice(),
                    item.getSellingPrice(),
                    item.isAvailable()
                );
                clonedItem.setSalesCount(item.getSalesCount());  // Копируем счетчик продаж
                cloned.menuItems.add(clonedItem);
            }
        }
        cloned.calculateTotalAmount();  // Пересчитываем сумму для глубокой копии
        return cloned;
    }
}

